File: js/src/core/DataPipelineClient.js
const axios = require('axios');
const { generateHmacSignature } = require('../utils/crypto');

/**
 * Data Pipeline API Client
 * Main client class for interacting with the Optikpi Data Pipeline API
 */
class DataPipelineClient {
  constructor(config = {}) {
    this.config = {
      baseURL: config.baseURL || 'https://demo.optikpi.com/apigw/ingest',
      authToken: config.authToken,
      accountId: config.accountId,
      workspaceId: config.workspaceId,
      timeout: config.timeout || 30000,
      retries: config.retries || 3,
      retryDelay: config.retryDelay || 1000,
      ...config
    };

    this.validateConfig();
    this.setupAxiosInstance();
  }

  /**
   * Validates the client configuration
   * @throws {Error} If required configuration is missing
   */
  validateConfig() {
    const { authToken, accountId, workspaceId } = this.config;
    
    if (!authToken) {
      throw new Error('authToken is required');
    }
    if (!accountId) {
      throw new Error('accountId is required');
    }
    if (!workspaceId) {
      throw new Error('workspaceId is required');
    }
  }

  /**
   * Sets up the axios instance with default configuration
   */
  setupAxiosInstance() {
    this.axios = axios.create({
      baseURL: this.config.baseURL,
      timeout: this.config.timeout,
      headers: {
        'Content-Type': 'application/json',
        'User-Agent': 'Optikpi-DataPipeline-SDK/1.0.0'
      }
    });

    // Add request interceptor for authentication
    this.axios.interceptors.request.use(
      (config) => {
        if (config.data && config.method !== 'get') {
          const hmacSignature = generateHmacSignature(
            config.data,
            this.config.authToken,
            this.config.accountId,
            this.config.workspaceId
          );

          config.headers['x-optikpi-token'] = this.config.authToken;
          config.headers['x-optikpi-account-id'] = this.config.accountId;
          config.headers['x-optikpi-workspace-id'] = this.config.workspaceId;
          config.headers['x-hmac-signature'] = hmacSignature;
          config.headers['x-hmac-algorithm'] = 'sha256';
        }
        return config;
      },
      (error) => {
        return Promise.reject(error);
      }
    );

    // Add response interceptor for error handling
    this.axios.interceptors.response.use(
      (response) => response,
      async (error) => {
        if (error.response && error.response.status >= 500 && this.config.retries > 0) {
          return this.retryRequest(error.config);
        }
        return Promise.reject(error);
      }
    );
  }

  /**
   * Retries a failed request
   * @param {Object} config - Axios request config
   * @returns {Promise} Axios response
   */
  async retryRequest(config, retryCount = 0) {
    if (retryCount >= this.config.retries) {
      throw new Error(`Request failed after ${this.config.retries} retries`);
    }

    await new Promise(resolve => setTimeout(resolve, this.config.retryDelay * (retryCount + 1)));
    
    try {
      return await this.axios.request(config);
    } catch (error) {
      if (error.response && error.response.status >= 500) {
        return this.retryRequest(config, retryCount + 1);
      }
      throw error;
    }
  }

  /**
   * Performs a health check on the API
   * @returns {Promise<Object>} Health check response
   */
  async healthCheck() {
    try {
      const response = await this.axios.get('/datapipeline/health');
      return {
        success: true,
        status: response.status,
        data: response.data,
        timestamp: new Date().toISOString()
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
        status: error.response?.status,
        timestamp: new Date().toISOString()
      };
    }
  }

  /**
   * Sends customer profile data
   * @param {Object|Array} data - Customer profile data or array of profiles
   * @returns {Promise<Object>} API response
   */
  async sendCustomerProfile(data) {
    try {
      const response = await this.axios.post('/customers', data);
      return {
        success: true,
        status: response.status,
        data: response.data,
        timestamp: new Date().toISOString()
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
        status: error.response?.status,
        data: error.response?.data,
        timestamp: new Date().toISOString()
      };
    }
  }

  /**
   * Sends account event data
   * @param {Object|Array} data - Account event data or array of events
   * @returns {Promise<Object>} API response
   */
  async sendAccountEvent(data) {
    try {
      const response = await this.axios.post('/events/account', data);
      return {
        success: true,
        status: response.status,
        data: response.data,
        timestamp: new Date().toISOString()
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
        status: error.response?.status,
        data: error.response?.data,
        timestamp: new Date().toISOString()
      };
    }
  }

  /**
   * Sends deposit event data
   * @param {Object|Array} data - Deposit event data or array of events
   * @returns {Promise<Object>} API response
   */
  async sendDepositEvent(data) {
    try {
      const response = await this.axios.post('/events/deposit', data);
      return {
        success: true,
        status: response.status,
        data: response.data,
        timestamp: new Date().toISOString()
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
        status: error.response?.status,
        data: error.response?.data,
        timestamp: new Date().toISOString()
      };
    }
  }

  /**
   * Sends withdrawal event data
   * @param {Object|Array} data - Withdrawal event data or array of events
   * @returns {Promise<Object>} API response
   */
  async sendWithdrawEvent(data) {
    try {
      const response = await this.axios.post('/events/withdraw', data);
      return {
        success: true,
        status: response.status,
        data: response.data,
        timestamp: new Date().toISOString()
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
        status: error.response?.status,
        data: error.response?.data,
        timestamp: new Date().toISOString()
      };
    }
  }

  /**
   * Sends gaming activity event data
   * @param {Object|Array} data - Gaming activity event data or array of events
   * @returns {Promise<Object>} API response
   */
  async sendGamingActivityEvent(data) {
    try {
      const response = await this.axios.post('/events/gaming-activity', data);
      return {
        success: true,
        status: response.status,
        data: response.data,
        timestamp: new Date().toISOString()
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
        status: error.response?.status,
        data: error.response?.data,
        timestamp: new Date().toISOString()
      };
    }
  }

  /**
   * Sends extended attributes data
   * @param {Object|Array} data - Extended attributes data or array of attributes
   * @returns {Promise<Object>} API response
   */
  async sendExtendedAttributes(data) {
    try {
      const response = await this.axios.post('/extattributes', data);
      return {
        success: true,
        status: response.status,
        data: response.data,
        timestamp: new Date().toISOString()
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
        status: error.response?.status,
        data: error.response?.data,
        timestamp: new Date().toISOString()
      };
    }
  }

  /**
   * Sends multiple events in batch
   * @param {Object} batchData - Object containing different event types
   * @returns {Promise<Object>} Batch response results
   */
  async sendBatch(batchData) {
    const results = {};
    const promises = [];

    if (batchData.customers) {
      promises.push(
        this.sendCustomerProfile(batchData.customers)
          .then(result => { results.customers = result; })
      );
    }

    if (batchData.accountEvents) {
      promises.push(
        this.sendAccountEvent(batchData.accountEvents)
          .then(result => { results.accountEvents = result; })
      );
    }

    if (batchData.depositEvents) {
      promises.push(
        this.sendDepositEvent(batchData.depositEvents)
          .then(result => { results.depositEvents = result; })
      );
    }

    if (batchData.withdrawEvents) {
      promises.push(
        this.sendWithdrawEvent(batchData.withdrawEvents)
          .then(result => { results.withdrawEvents = result; })
      );
    }

    if (batchData.gamingEvents) {
      promises.push(
        this.sendGamingActivityEvent(batchData.gamingEvents)
          .then(result => { results.gamingEvents = result; })
      );
    }

    if (batchData.extendedAttributes) {
      promises.push(
        this.sendExtendedAttributes(batchData.extendedAttributes)
          .then(result => { results.extendedAttributes = result; })
      );
    }

    await Promise.all(promises);

    return {
      success: true,
      results,
      timestamp: new Date().toISOString()
    };
  }

  /**
   * Updates client configuration
   * @param {Object} newConfig - New configuration options
   */
  updateConfig(newConfig) {
    this.config = { ...this.config, ...newConfig };
    this.validateConfig();
    this.setupAxiosInstance();
  }

  /**
   * Gets current configuration
   * @returns {Object} Current configuration (without sensitive data)
   */
  getConfig() {
    const { authToken, ...safeConfig } = this.config;
    return {
      ...safeConfig,
      authToken: authToken ? `${authToken.substring(0, 8)}...` : undefined
    };
  }
}

module.exports = DataPipelineClient;

File: python/src/python/core/DataPipelineClient.py
"""
Data Pipeline API Client
Main client class for interacting with the Optikpi Data Pipeline API
"""
import time
import json
from typing import Dict, Any, List, Union, Optional
from datetime import datetime
import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry

from utils.crypto import generate_hmac_signature


class DataPipelineClient:
    """
    Data Pipeline API Client
    
    Main client class for interacting with the Optikpi Data Pipeline API.
    Handles authentication, retries, and all API endpoints.
    """
    
    def __init__(self, config: Optional[Dict[str, Any]] = None):
        """
        Initialize the Data Pipeline Client
        
        Args:
            config: Configuration dictionary with:
                - baseURL: API base URL (required)
                - authToken: Authentication token (required)
                - accountId: Account ID (required)
                - workspaceId: Workspace ID (required)
                - timeout: Request timeout in seconds (default: 30)
                - retries: Number of retries (default: 3)
                - retryDelay: Retry delay in seconds (default: 1)
        """
        if config is None:
            config = {}
            
        self.config = {
            'baseURL': config.get('baseURL', 'https://demo.optikpi.com/apigw/ingest'),
            'authToken': config.get('authToken'),
            'accountId': config.get('accountId'),
            'workspaceId': config.get('workspaceId'),
            'timeout': config.get('timeout', 30),
            'retries': config.get('retries', 3),
            'retryDelay': config.get('retryDelay', 1),
            **config
        }
        
        self.validate_config()
        self.setup_session()
    
    def validate_config(self) -> None:
        """
        Validates the client configuration
        
        Raises:
            ValueError: If required configuration is missing
        """
        auth_token = self.config.get('authToken')
        account_id = self.config.get('accountId')
        workspace_id = self.config.get('workspaceId')
        
        if not auth_token:
            raise ValueError('authToken is required')
        if not account_id:
            raise ValueError('accountId is required')
        if not workspace_id:
            raise ValueError('workspaceId is required')
    
    def setup_session(self) -> None:
        """
        Sets up the requests session with default configuration and retry logic
        """
        self.session = requests.Session()
        
        # Set up retry strategy
        retry_strategy = Retry(
            total=self.config['retries'],
            backoff_factor=self.config['retryDelay'],
            status_forcelist=[500, 502, 503, 504],
            allowed_methods=["POST", "GET"]
        )
        
        adapter = HTTPAdapter(max_retries=retry_strategy)
        self.session.mount("http://", adapter)
        self.session.mount("https://", adapter)
        
        # Set default headers
        self.session.headers.update({
            'Content-Type': 'application/json',
            'User-Agent': 'Optikpi-DataPipeline-SDK-Python/1.0.0'
        })
    
    def _add_auth_headers(self, data: Any) -> Dict[str, str]:
        """
        Generate authentication headers including HMAC signature
        
        Args:
            data: Request data to sign
            
        Returns:
            Dictionary of authentication headers
        """
        hmac_signature = generate_hmac_signature(
            data,
            self.config['authToken'],
            self.config['accountId'],
            self.config['workspaceId']
        )
        
        return {
            'x-optikpi-token': self.config['authToken'],
            'x-optikpi-account-id': self.config['accountId'],
            'x-optikpi-workspace-id': self.config['workspaceId'],
            'x-hmac-signature': hmac_signature,
            'x-hmac-algorithm': 'sha256'
        }
    
    def _make_request(self, method: str, endpoint: str, data: Any = None) -> Dict[str, Any]:
        """
        Make an HTTP request to the API
        
        Args:
            method: HTTP method (GET, POST, etc.)
            endpoint: API endpoint path
            data: Request data
            
        Returns:
            Response dictionary with success status and data
        """
        url = f"{self.config['baseURL']}{endpoint}"
        
        try:
            headers = {}
            request_body = None
            
            if method != 'GET' and data is not None:
                # CRITICAL FIX: Generate signature from data object first
                headers = self._add_auth_headers(data)
                
                # Then convert to JSON string for the request body
                # Use same format as signature: separators=(',', ':')
                request_body = json.dumps(data, separators=(',', ':'))
            
            # Make the request with pre-serialized JSON string
            response = self.session.request(
                method=method,
                url=url,
                data=request_body,  # Send as string, not json=data
                headers=headers,
                timeout=self.config['timeout']
            )
            
            # Parse response
            try:
                response_data = response.json()
            except ValueError:
                response_data = {'message': response.text} if response.text else {}
            
            if response.ok:
                return {
                    'success': True,
                    'status': response.status_code,
                    'data': response_data,
                    'timestamp': datetime.now().isoformat()
                }
            else:
                return {
                    'success': False,
                    'status': response.status_code,
                    'error': response_data.get('message', response.text),
                    'data': response_data,
                    'timestamp': datetime.now().isoformat()
                }
        
        except requests.exceptions.Timeout:
            return {
                'success': False,
                'error': f'Request timeout after {self.config["timeout"]} seconds',
                'status': 0,
                'timestamp': datetime.now().isoformat()
            }
        except requests.exceptions.ConnectionError as e:
            return {
                'success': False,
                'error': f'Connection error: {str(e)}',
                'status': 0,
                'timestamp': datetime.now().isoformat()
            }
        except Exception as e:
            return {
                'success': False,
                'error': str(e),
                'status': 0,
                'timestamp': datetime.now().isoformat()
            }
    
    def send_customer_profile(self, data: Union[Dict[str, Any], List[Dict[str, Any]]]) -> Dict[str, Any]:
        """
        Sends customer profile data
        
        Args:
            data: Customer profile data or list of profiles
            
        Returns:
            API response dictionary
        """
        return self._make_request('POST', '/customers', data)
    
    def send_account_event(self, data: Union[Dict[str, Any], List[Dict[str, Any]]]) -> Dict[str, Any]:
        """
        Sends account event data
        
        Args:
            data: Account event data or list of events
            
        Returns:
            API response dictionary
        """
        return self._make_request('POST', '/events/account', data)
    
    def send_deposit_event(self, data: Union[Dict[str, Any], List[Dict[str, Any]]]) -> Dict[str, Any]:
        """
        Sends deposit event data
        
        Args:
            data: Deposit event data or list of events
            
        Returns:
            API response dictionary
        """
        return self._make_request('POST', '/events/deposit', data)
    
    def send_withdraw_event(self, data: Union[Dict[str, Any], List[Dict[str, Any]]]) -> Dict[str, Any]:
        """
        Sends withdrawal event data
        
        Args:
            data: Withdrawal event data or list of events
            
        Returns:
            API response dictionary
        """
        return self._make_request('POST', '/events/withdraw', data)
    
    def send_gaming_activity_event(self, data: Union[Dict[str, Any], List[Dict[str, Any]]]) -> Dict[str, Any]:
        """
        Sends gaming activity event data
        
        Args:
            data: Gaming activity event data or list of events
            
        Returns:
            API response dictionary
        """
        return self._make_request('POST', '/events/gaming-activity', data)

    def send_extended_attributes(self, data: Union[Dict[str, Any], List[Dict[str, Any]]]) -> Dict[str, Any]:
        """
        send Extended Attributes data
        
        Args:
            data: send Extended Attributes data or list of events
            
        Returns:
            API response dictionary
        """
        return self._make_request('POST', '/extattributes', data)
    
    def send_wallet_balance_event(self, data: Union[Dict[str, Any], List[Dict[str, Any]]]) -> Dict[str, Any]:
        """
        Sends wallet balance event data
    
        Args:
            data: Wallet balance event data or list of events
        
        Returns:
            API response dictionary
        """
        return self._make_request('POST', '/events/wallet-balance', data)


    def send_refer_friend_event(self, data: Union[Dict[str, Any], List[Dict[str, Any]]]) -> Dict[str, Any]:
        """
        Sends refer friend event data
    
        Args:
            data: Refer friend event data or list of events
        
        Returns:
            API response dictionary
        """
        return self._make_request('POST', '/events/refer-friend', data)

    
    def send_batch(self, batch_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Sends multiple events in batch
        
        Args:
            batch_data: Dictionary containing different event types:
                - customers: List of customer profiles
                - accountEvents: List of account events
                - depositEvents: List of deposit events
                - withdrawEvents: List of withdraw events
                - gamingEvents: List of gaming activity events
                - walletBalanceEvents: List of wallet balance events
                - referFriendEvents: List of refer friend events
                - customerExtEvents: List of customer extended attributes events
            
        Returns:
            Batch response results dictionary
        """
        results = {}
        
        # Send each type of data
        if 'customers' in batch_data:
            results['customers'] = self.send_customer_profile(batch_data['customers'])
        
        if 'accountEvents' in batch_data:
            results['accountEvents'] = self.send_account_event(batch_data['accountEvents'])
        
        if 'depositEvents' in batch_data:
            results['depositEvents'] = self.send_deposit_event(batch_data['depositEvents'])
        
        if 'withdrawEvents' in batch_data:
            results['withdrawEvents'] = self.send_withdraw_event(batch_data['withdrawEvents'])
        
        if 'gamingEvents' in batch_data:
            results['gamingEvents'] = self.send_gaming_activity_event(batch_data['gamingEvents'])
        
        if 'walletBalanceEvents' in batch_data:
             results['walletBalanceEvents'] = self.send_wallet_balance_event(batch_data['walletBalanceEvents'])

        if 'referFriendEvents' in batch_data:
            results['referFriendEvents'] = self.send_refer_friend_event(batch_data['referFriendEvents'])

        if 'extendedAttributes' in batch_data:
            results['extendedAttributes'] = self.send_extended_attributes(batch_data['extendedAttributes'])
    
        


        return {
            'success': True,
            'results': results,
            'timestamp': datetime.now().isoformat()
        }
    
    def update_config(self, new_config: Dict[str, Any]) -> None:
        """
        Updates client configuration
        
        Args:
            new_config: Dictionary with new configuration options
        """
        self.config.update(new_config)
        self.validate_config()
        self.setup_session()
    
    
    def get_config(self) -> Dict[str, Any]:
        """
        Gets current configuration
        
        Returns:
            Current configuration (without sensitive data)
        """
        safe_config = self.config.copy()
        auth_token = safe_config.get('authToken', '')
        if auth_token:
            safe_config['authToken'] = f"{auth_token[:8]}..."
        return safe_config
    
    def __enter__(self):
        """Context manager entry"""
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit"""
        self.session.close()
    
    def close(self):
        """Close the session"""
        self.session.close()